### PR TITLE
Implement a Reader

### PR DESCRIPTION
**Pull request by Keksoj**
The Apache Software Foundation considers integrating pulsar-rs, see issue #126 

In order to expand the rust client's features, we need to implement a Reader interface, as suggested at issue #144 . A Reader is a pulsar client that immediately aknowledges messages upon reception.

### CODE REVIEW

**Review by Geal — CHANGES_REQUESTED** [review]

**Review by Geal — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 .idea
 **/*.rs.bk
 Cargo.lock
+.vscode/
```
Comment (Geal) [inline .gitignore]: this should be removed (we won't start adding special case for every editor)

File: README.md [hunk 2]
```diff
@@ -1,6 +1,8 @@
 ## Pulsar
 #### Future-based Rust client for [Apache Pulsar](https://pulsar.apache.org/)
 
+![under construction](http://www.chipenergy.com/images/Under_Construction.gif)
```
Comment (Geal) [inline README.md]: this should be removed

File: examples/reader.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        builder = builder.with_auth(authentication);
+    }
+
+    let pulsar: Pulsar<_> = builder.build().await?;
+
+    let mut reader: Reader<TestData, _> = pulsar
+        .reader()
+        .with_topic(topic)
+        .with_consumer_name("test_reader")
+        .with_options(
```
Comment (Geal) [inline examples/reader.rs]: why is the reader example setting the durable option and the initial position? Those should be set already when creating the reader(durable to false, initial position to latest)

File: src/client.rs [hunk 4]
```diff
@@ -271,6 +272,15 @@ impl<Exe: Executor> Pulsar<Exe> {
         ProducerBuilder::new(self)
     }
 
+    /// creates a reader builder
+    /// ```rust, no_run
+    /// // please fill out an example
```
Comment (Geal) [inline src/client.rs]: please fill out an example :wink:

File: src/client.rs [hunk 5]
```diff
@@ -271,6 +272,15 @@ impl<Exe: Executor> Pulsar<Exe> {
...
+    /// ```
+    pub fn reader(&self) -> ConsumerBuilder<Exe> {
+        // this makes it exactly the same like the consumer() method though
+        ConsumerBuilder::new(self)
```
Comment (Geal) [inline src/client.rs]: this is where the durable option and the initial position should be set

File: src/consumer.rs [hunk 6]
```diff
@@ -61,6 +62,44 @@ pub struct ConsumerOptions {
     pub initial_position: InitialPosition,
 }
 
+impl ConsumerOptions {
```
Comment (Geal) [inline src/consumer.rs:65]: those methods could be proposed in a separate pull request

File: src/consumer.rs [hunk 7]
```diff
@@ -61,6 +62,44 @@ pub struct ConsumerOptions {
...
+    /// within options, sets the priority level
+    pub fn with_priority_level(mut self, priority_level: i32) -> Self {
+        self.priority_level = Some(priority_level);
+        self
+    }
+
+    pub fn with_durable_option(mut self, durable: bool) -> Self {
```
Comment (Geal) [inline src/consumer.rs]: we don't need the `_option` suffix

File: src/consumer.rs [hunk 8]
```diff
@@ -1454,10 +1501,9 @@ impl<Exe: Executor> ConsumerBuilder<Exe> {
             consumer_name,
             batch_size,
             unacked_message_resend_delay,
-            namespace,
-            topic_refresh,
             consumer_options,
             dead_letter_policy,
+            ..
```
Comment (Geal) [inline src/consumer.rs]: if there are more fields in the builder, they should be used, not silently ignored

File: src/consumer.rs [hunk 9]
```diff
@@ -372,17 +411,18 @@ enum InnerConsumer<T: DeserializeMessage, Exe: Executor> {
...
 type MessageIdDataReceiver = mpsc::Receiver<Result<(proto::MessageIdData, Payload), Error>>;
 
+// this is entirely public for use in reader.rs
 pub(crate) struct TopicConsumer<T: DeserializeMessage, Exe: Executor> {
-    consumer_id: u64,
-    config: ConsumerConfig,
-    topic: String,
-    messages: Pin<Box<MessageIdDataReceiver>>,
-    engine_tx: mpsc::UnboundedSender<EngineMessage<Exe>>,
+    pub(crate) consumer_id: u64,
```
Comment (Geal) [inline src/consumer.rs:416]: do we need all of those to be pub(crate)?

**Review by Geal — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (Keksoj) [reply .gitignore]: done

File: README.md (hunk 2 above)
Comment (Keksoj) [reply README.md]: done

File: examples/reader.rs (hunk 3 above)
Comment (Keksoj) [reply examples/reader.rs]: sure, you're right. it's done

File: src/client.rs (hunk 5 above)
Comment (Keksoj) [reply src/client.rs]: done.

File: src/client.rs (hunk 4 above)
Comment (Keksoj) [reply src/client.rs]: done!

File: src/consumer.rs (hunk 6 above)
Comment (Keksoj) [reply src/consumer.rs:65]: That may be true, but they are usefull in the reader examples, while remaining optional and leaving backward compatibility. It would be as practical to just deliver them in this PR.

Comment (Geal) [reply src/consumer.rs:65]: ok

File: src/consumer.rs (hunk 9 above)
Comment (Keksoj) [reply src/consumer.rs:416]: as a matter of fact, no, we don't. I ran `cargo check`ed all of them, only `consumer_id`, `config`, and `dead_letter_policy` are needed elsewhere.

File: src/consumer.rs (hunk 8 above)
Comment (Keksoj) [reply src/consumer.rs]: done with `unused_variables` macro

**Comment by Geal** [comment]
thanks!
